rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,554 @@
+
+Its goal is to be invoked like so:
+
+```rust
+rsx! {
+  <Element>
+  	{
+    	  MyManualWidget::new(CustomFoo(123))
+  	}
+  </Element>
+}
+```
+
+To do this, it has to implement `Widget`:
+
+```rust
+impl Widget for MyManualWidget {
+  fn get_id(&self) -> Index { self.my_id }
+  // ...
+  fn render(&mut self, mut context: KayakContextRef) {
+    // ???
+  }
+}
+```
+
+Currently, the only way to build this widget is to do something like:
+
+```rust
+// fragment.rs
+
+fn render(&mut self, context: &mut KayakContext) {
+  let parent_id = self.get_id();
+  let tree = crate::WidgetTree::new();
+  tree.add(parent_id, None);
+
+  if let Some(children) = self.children.take() {
+      children(tree.clone(), Some(parent_id), context);
+  } else {
+      return;
+  }
+
+  // Consume the widget tree taking the inner value
+  let tree = tree.take();
+
+  // Evaluate changes to the tree.
+  let changes = context.widget_manager.tree.diff_children(&tree, parent_id);
+
+  context.widget_manager.tree.merge(&tree, parent_id, changes);
+}
+```
+
+But without access to `KayakContext` we can't quite do this anymore (which is good because it's a lot of boilerplate to write).
+
+There are two options for handling this: move logic into `KayakContextRef` or create a `WidgetTreeBuilder`.
+
+##### Extending KayakContextRef
+
+This would add the following methods:
+
+```rust
+pub fn create_tree(&self) -> WidgetTree {
+  let tree = WidgetTree::new();
+  tree.add(self.current_id, None);
+  tree
+}
+
+// Pretend `Children` isn't wrapped by `Option<...>`
+pub fn add_children(&mut self, children: Children, tree: &WidgetTree) {
+	let id =  self.current_id;
+	children(tree.clone(), Some(id), context);
+}
+
+pub fn add_widget<W: Widget>(&mut self, widget: W, widget_index: usize, tree: &WidgetTree) {
+  let id =  self.current_id;
+  let (should_rerender, child_id) = self.context
+    .widget_manager
+    .create_widget(widget_index, widget.clone(), Some(id));
+  tree.add(child_id, Some(id));
+  if should_rerender {
+      let mut child_widget = self.context.widget_manager.take(child_id);
+      let mut context = KayakContextRef { context: self.context, current_id: id };
+      child_widget.render(context);
+      self.context.widget_manager.repossess(child_widget);
+  }
+}
+
+pub fn commit_tree(&mut self, tree: WidgetTree) {
+  let tree = tree.take();
+  let id =  self.current_id;
+  let changes = self.context.widget_manager.tree.diff_children(&tree, id);
+  self.context.widget_manager.tree.merge(&tree, id, changes);
+}
+```
+
+And results in usage like so:
+
+```rust
+// fragment.rs
+
+fn render(&mut self, mut context: KayakContextRef) {
+	let tree = context.create_tree();
+  if let Some(children) = self.children.take() {
+    context.add_children(children, &tree);
+  } else {
+      return;
+  }
+  context.commit_tree(tree);
```

Comment:
I think we can just call this `commit` perhaps?

Reply:
I named it to be a bit more clear amongst its other methods (`bind`, `create_state`, etc.). But we can go ahead and just call it `commit` I guess since it only takes a tree anyways haha